source/agora/common/Amount.d

```diff
@@ -0,0 +1,239 @@
+    }
+
+
+    nothrow pure @nogc @safe:
+
+    /***************************************************************************
+
+        Construct an instance of an `Amount`
+
+        Params:
+            units = The monetary amount this new instance represent.
+                    The unit is the smaller unit of currency (see `UnitPerCoin`)
+
+    ***************************************************************************/
+
+    public this (ulong units)
+    {
+        assert(Amount.isInRange(units));
+        this.value = units;
+    }
+
+    /// Make a value without checking bounds
+    /// Used to initialize the bounds themselves, and to make invalid values
+    /// in unittests
+    private this (ulong units, bool dummy) { this.value = units; }
+    /// Ditto
+    version (unittest)
+    public static Amount invalid (ulong units) { return Amount(units, true); }
+
+    /***************************************************************************
+
+        Returns:
+            `true` if `value` is in the range `[0; Amount.MaxUnitSupply]`
+            and can be safely used to make an `Amount`
+
+    ***************************************************************************/
+
+    pragma(inline, true)
+    public static bool isInRange (ulong value)
+    {
+        return  value <= MaxUnitSupply.value;
+    }
+
+    /// `isInRange` but as member function
+    pragma(inline, true)
+    public bool isValid () const
+    {
+        return isInRange(this.value);
+    }
+
+
+    /***************************************************************************
+
+        Add another value to `this` `Amount`
+
+        Params:
+            other = Another `Amount` value. If the value is invalid,
+                    `false` will be returned.
+        Returns:
+            `true` if the addition returned a number within bounds.
+            `false` if the number is out of the [0; MaxUnitSupply] bound.
+
+    ***************************************************************************/
+
+    pragma(inline, true)
+    public bool add (Amount other)
+    {
+        bool overflow;
+        this.value = addu(this.value, other.value, overflow);
+        if (overflow || !this.isValid())
+        {
+            // If we overflow, make sure to poison the return value
+            this.value = ulong.max;
+            return false;
+        }
+        return true;
+    }
+
+    /***************************************************************************
+
+        Substract another value from `this` `Amount`
+
+        Params:
+            other = Another `Amount` value. If the value is invalid,
+                    `false` will be returned.
+        Returns:
+            `true` if the addition returned a number within bounds.
+            `false` if the number is out of the [0; MaxUnitSupply] bound.
+
+    ***************************************************************************/
+
+    pragma(inline, true)
+    public bool sub (Amount other)
+    {
```

Comment:
Hmm.. What about checking if `this.isValid` is false at the start of this function? Because the value may have already be poisoned.

Reply:
And the subtracting it "unpoisons" it by mistake.

Reply:
Very good point. Fixed and added a test that substracts `invalid` to `invalid`.